### PR TITLE
Prepare to switch from dnf to rpm in our pipelines

### PR DESCRIPTION
**Pull request by teg**
These preliminary commits adds the needed low-level features to be able to use the rpm rather than the dnf stage. The switchover to change the produced pipelines is not yet done, so this can be merged independently of the required changes in osbuild.

### CODE REVIEW

File: internal/weldr/api.go [hunk 1]
```diff
@@ -1757,7 +1757,14 @@ func (api *API) fetchPackageList() (rpmmd.PackageList, error) {
 	return packages, err
 }
 
-func (api *API) depsolveBlueprint(bp *blueprint.Blueprint, clean bool) ([]rpmmd.PackageSpec, map[string]string, error) {
+func (api *API) depsolveBlueprint(bp *blueprint.Blueprint, outputType, arch string, clean bool) ([]rpmmd.PackageSpec, []rpmmd.PackageSpec, map[string]string, error) {
```
Comment (msehnout) [inline internal/weldr/api.go]: I'd prefer if new code used the types from `internal/common` and variable names reflected what is written in the README file. So I'd suggest to replace `outputType` with `imageType`.

File: internal/weldr/api.go [hunk 2]
```diff
@@ -1769,16 +1776,33 @@ func (api *API) depsolveBlueprint(bp *blueprint.Blueprint, clean bool) ([]rpmmd.
 			specs[i] += "-*-*.*"
 		}
 	}
+	excludeSpecs := []string{}
+	if outputType != "" {
```
Comment (msehnout) [inline internal/weldr/api.go]: How is it possible to reach this function without an image type? I thought depsolv without the set of base packages does not make any sense.

File: internal/weldr/api.go [hunk 3]
```diff
@@ -1769,16 +1776,33 @@ func (api *API) depsolveBlueprint(bp *blueprint.Blueprint, clean bool) ([]rpmmd.
...
+		return nil, nil, nil, err
 	}
-	for _, source := range api.store.GetAllSources() {
-		repos = append(repos, source.RepoConfig())
+
+	buildPackages := []rpmmd.PackageSpec{}
+	if outputType != "" {
+		buildSpecs, err := api.distro.BuildPackages(arch)
+		if err != nil {
+			return nil, nil, nil, err
```
Comment (msehnout) [inline internal/weldr/api.go]: Do we have a reasonable way to tell the user what happened here? Was it possible to have such errors in Weldr?

File: internal/osbuild/source.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+		return err
+	}
+	*sources = make(map[string]Source)
+	for name, rawSource := range rawSources {
+		var source Source
+		switch name {
+		case "org.osbuild.files":
+			source = new(FilesSource)
+		default:
+			return errors.New("unexpected suorce name" + name)
```
Comment (msehnout) [inline internal/osbuild/source.go]: Just a small typo :)

File: internal/distro/fedora30/distro.go [hunk 5]
```diff
@@ -436,16 +455,11 @@ func (r *Fedora30) Runner() string {
...
-		"e2fsprogs",
-		"policycoreutils",
-		"qemu-img",
-		"systemd",
-		"tar",
-	}
-	packages = append(packages, arch.BuildPackages...)
+	packages, err := r.BuildPackages(arch.Name)
+	if err != nil {
+		panic("impossibly invalid arch")
```
[the remark this answers is not in the document]
Comment (jkozol) [reply internal/distro/fedora30/distro.go]: I agree that this is not a widely used phrase and "invalid arch" or "impossible and invalid arch" might be technically better but I think it explains the error fine and I personally enjoy the phrase "impossibly invalid arch".

[the remark this answers is not in the document]
Comment (teg) [reply internal/distro/fedora30/distro.go]: I spent way too long trying to decide what to do with this feedback ;) In the end: yes, the phrase has a nice sound to it, but I'll make it grammatically correct.

File: internal/weldr/api.go (hunk 2 above)
Comment (teg) [reply internal/weldr/api.go]: I decided to leave this in for now, and rather clean up all the naming in one commit/PR.

File: internal/weldr/api.go (hunk 3 above)
Comment (teg) [reply internal/weldr/api.go]: This would be an programming error on our side. We should `panic()` or at most return `InternalServerError`.

**Comment by teg** [comment]
Rebased on master, fixed according to the reviews and dropped the final commit that changes the output of osbuild-pipeline. This should be done after osbuild itself has been adjusted.

File: internal/distro/test/distro.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
+package test
```
Comment (msehnout) [inline internal/distro/test/distro.go]: Is this package used anywhere? I replaced all usages with the fedoratest packages if I remember correctly.

File: internal/osbuild/source.go (hunk 4 above)
Comment (msehnout) [reply internal/osbuild/source.go]: Still there ;-)

**Comment by msehnout** [comment]
As other were fine with this PR already, I'm going to merge this now. Thanks @teg !

**Pull request merged**